cmd/admin-heal-ops.go

```diff
@@ -519,6 +525,67 @@ func (h *healSequence) healSequenceStart() {
 	}
 }
 
+func (h *healSequence) queueHealTask(path string, healType madmin.HealItemType) error {
+	var respCh = make(chan healResult)
+	defer close(respCh)
+	// Send heal request
+	globalBackgroundHealing.queueHealTask(healTask{path: path, responseCh: respCh, opts: h.settings})
+	// Wait for answer and push result to the client
+	res := <-respCh
+	if !h.reportProgress {
+		return nil
+	}
+	res.result.Type = healType
+	if res.err != nil {
+		// Object might have been deleted, by the time heal
+		// was attempted we ignore this object an move on.
+		if isErrObjectNotFound(res.err) {
+			return nil
+		}
+		// Only report object error
+		if healType != madmin.HealItemObject {
+			return res.err
+		}
+		res.result.Detail = res.err.Error()
+	}
+	return h.pushHealResultItem(res.result)
+}
+
+func (h *healSequence) healFromSourceCh() {
+	var err error
+	checkErr := func(f func() error) {
+		switch {
+		case err != nil:
```

Comment:
There is no race/shadow problem here. The idea is to stop calling heal functions after an unrecoverable error has happened. Instead of repetitively writing an error check block like:

```go
err = doSomething()
if err != nil {
     h.traverseAndHealDoneCh <- err
     return
} else if h.isQuitting() {
     h.traverseAndHealDoneCh <- errHealStopSignalled
     return
}
// Repeat all of the above now for doSomething2() and so on.
```

the checks are distilled into a simple closure. It is an idiomatic style: https://blog.golang.org/errors-are-values

Reply:
I disagree, we can make this part of code more testable, than stacking errors in this manner. The race may have been the wrong word to use here. It is more about code simplicity. 

This style of coding is clearly not simple, it is each functions job to short circuit itself other than relying on a previous error cache. 

The simplest is this, more testable since now we have healAll() as a function which returns an error. Writing tests for closure based functions is very hard in practice. 

```go
func (h *healSequence) healFromSourceCh() {
        defer close(h.traverseAndHealDoneCh)

        healAll := func() error {
                // Start with format healing
                if err := h.healDiskFormat(); err != nil {
                        return err
                }

                // Start healing the config prefix.
                if err := h.healMinioSysMeta(minioConfigPrefix)(); err != nil {
                        return err
                }

                // Start healing the bucket config prefix.
                if err := h.healMinioSysMeta(bucketConfigPrefix)(); err != nil {
                        return err
                }

                for path := range h.sourceCh {
                        if err := h.queueHealTask(path, madmin.HealItemObject); err != nil {
                                return err
                        }
                }

                return nil
        }

        if err := healAll(); err != nil {
                h.traverseAndHealDoneCh <- err
        }
}
```

In the same article that you reference, it is preferred to not use closure, to be cleaner and that is my intention here and to be more testable.

Reply:
IMO we should rely on Go's tricks rather write straight forward code, a problem of closure as shown here can clearly show what is going on.

```go
package main

import (
	"errors"
	"fmt"
)

func main() {
	var err error
	checkErr := func(f func() error) {
		switch {
		case err != nil:
			return
		}

		err := f()
		if err != nil {
			fmt.Println(err)
			return
		}
	}

	checkErr(func() error { return errors.New("hello") })
	fmt.Println(err)
}
```

The problem is Go doesn't complain about such issues and one has to be really careful about making sure that variables are not re-defined and re-referenced.  This is a very tricky situation to be in, coupled with the value by reference kind of bugs are really harder to fix.

So the simplest way is to use what I wrote, avoids bugs and makes it more maintainable in the long run.

Reply:
Your version of the function does not check if the heal sequence is already quitting (perhaps because of a user signal, etc) before calling each heal operation. Without that there cannot be a force stop functionality. 

Also I don't think your version comes out clearly as simpler: (1) it does not do the same things as the original, (2) you can further simplify it by avoiding the defer statement (it's unnecessary - can just place the close at end of the function), (3) its not obvious how a test is more easy to write for this version than the first.

Reply:
The recommended way in the errors blog is to write a type that tracks its error state in-situ. Once an error happens, the function calls become no-op rather similar to how the existing code works. That does seem cleaner because the error checking, and checking for quitting status can be done in a single place that is not as distracting as a closure.

Reply:
As discussed in person with @harshavardhana - it is better to separate out notifying the heal error into a separate function, so that a test can be written to check if the internal heal routine quits on error, sort of like:

```go
func (h *healSequence) healItems() error {
	// Start with format healing
	if err := h.healDiskFormat(); err != nil {
		return err
	}

	// Start healing the config prefix.
	if err := h.healMinioSysMeta(minioConfigPrefix)(); err != nil {
		return err
	}

	// Start healing the bucket config prefix.
	if err := h.healMinioSysMeta(bucketConfigPrefix)(); err != nil {
		return err
	}

	for path := range h.sourceCh {
		if err := h.queueHealTask(path, madmin.HealItemObject); err != nil {
			return err
		}
	}

	return nil
}

func (h *healSequence) healFromSourceCh() {
	if err := h.healItems(); err != nil {
		h.traverseAndHealDoneCh <- err
	}
	close(h.traverseAndHealDoneCh)
}
```